src/React/Basic/Hooks/Internal.purs

```diff
@@ -0,0 +1,127 @@
+module React.Basic.Hooks.Internal
+  ( Render
+  , newtypeHook
+  , unsafeHook
+  , unsafeRenderEffect
+  , Pure
+  , Hook
+  , bind
+  , discard
+  ) where
+
+import Prelude hiding (bind)
+import Control.Applicative.Indexed (class IxApplicative)
+import Control.Apply.Indexed (class IxApply)
+import Control.Bind.Indexed (class IxBind, ibind)
+import Data.Functor.Indexed (class IxFunctor)
+import Data.Newtype (class Newtype)
+import Effect (Effect)
+import Prelude (bind) as Prelude
+import Type.Equality (class TypeEquals)
+
+-- | Render represents the effects allowed within a React component's
+-- | body, i.e. during "render". This includes hooks and ends with
+-- | returning JSX (see `pure`), but does not allow arbitrary side
+-- | effects.
+newtype Render x y a
+  = Render (Effect a)
+
+-- | Alias a chain of hooks. Useful for exposing a single "clean"
+-- | type when creating a hook to improve error messages
+-- | and hide implementation details.
+-- |
+-- | For example, the following alias is technically correct but
+-- | when inspecting types or error messages the alias is expanded
+-- | to the full original type:
+-- |
+-- | ```purs
+-- | type UseNodeDistance hooks = UseEffect Unit (UseState Int (UseRef (Nullable Node) hooks))
+-- |
+-- | useNodeDistanceFromMouse :: Hook UseNodeDistance (Int /\ (Ref (Nullable Node)))
+-- | ```
+-- |
+-- | `aliasHook` allows TODO!
+-- | when inspecting types or error messages the alias is expanded
+-- | to the full original type:
+-- |
+-- | ```purs
+-- | type UseNodeDistance hooks = UseEffect Unit (UseState Int (UseRef (Nullable Node) hooks))
+-- |
+-- | useNodeDistanceFromMouse :: Hook UseNodeDistance (Int /\ (Ref (Nullable Node)))
+-- | ```
+-- |
+-- |
+-- |
+newtypeHook ::
+  forall hooks oldHook newHook a.
+  Newtype newHook oldHook =>
+  Render hooks oldHook a ->
+  Render hooks newHook a
+newtypeHook (Render a) = Render a
```

Comment:
I thought about just calling this `[rename/alias]Hook` instead, but the newtype felt slightly more safe, maybe.. I'm actually not sure it's necessary at all for safety. Maybe I'm missing something.. the compiler just needs to infer the `oldHook` type as something that makes sense, it doesn't actually need to be a particular type. This does feel more principled though, because without the newtype constraint you could substitute any random type.. dunno. This is my biggest question in this review I think.

Reply:
I think a better for this could be `coerceHook`. Also, it looks like compiler support for `Coercible` is coming soon (see purescript#3351), so you'll probably want to replace this `Newtype` constraint with `Coercible` soon :)

I believe `Coercible` should be safe to use here anyway, as all the base hooks are foreign imports.